# Hoja de Trabajo 7
## Modelo SVM
---
**Integrantes**
- Maria Marta Ramirez Gil
- Gustavo Andres Gonzalez Pineda 
- Diego Alberto Leiva Pérez 
---

In [27]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

## Cargar los Datos

In [10]:
# Cargar los datos
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')

## Explorar los datos y hacer transformaciones

In [11]:
# Mostrar las columnas de cada archivo
print("\nColumns for train.csv")
print(df_test.columns)
print("\nColumns for test.csv")
print(df_train.columns)


Columns for train.csv
Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'G

In [12]:
# Mostar la forma de cada DataFrame
print("\nShape of DataFrame for train.csv")
print(df_test.shape)
print("\nShape of DataFrame for test.csv")
print(df_train.shape)


Shape of DataFrame for train.csv
(1459, 80)

Shape of DataFrame for test.csv
(1460, 81)


In [14]:
# Explorar los datos
print("Información de df_train:")
print(df_train.info())
print("Estadísticas de df_train:")
print(df_train.describe())

print("Información de df_test:")
print(df_test.info())
print("Estadísticas de df_test:")
print(df_test.describe())

Información de df_train:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-nul

### Nulos

In [18]:
print("Valores nulos en df_train:")
df_train.isnull().sum()

Valores nulos en df_train:


Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

In [19]:
print("Valores nulos en df_test:")
df_test.isnull().sum()

Valores nulos en df_test:


Id                 0
MSSubClass         0
MSZoning           4
LotFrontage      227
LotArea            0
                ... 
MiscVal            0
MoSold             0
YrSold             0
SaleType           1
SaleCondition      0
Length: 80, dtype: int64

### Transformacion
Para lograr nuestro objetivo de clasificar las casas en categorías de precios bajos, medios y altos, hemos introducido una nueva columna que simplifica este proceso. Hemos definido nuestras categorías de precios como baratos, medios y altos, basados en criterios propios.

In [21]:
def categorize_price(price):
    if price < 150000:
        return 'Cheap'
    elif price < 300000:
        return 'Medium'
    else:
        return 'Expensive'

# Aplicar la función para crear la nueva columna
df_train['PriceCategory'] = df_train['SalePrice'].apply(categorize_price)


In [22]:
# Verificar la creacion de la columna
print("Columnas de df_train:")
print(df_train.columns)

print("Valores únicos de la columna PriceCategory:")
print(df_train['PriceCategory'].unique())

Columnas de df_train:
Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'Ga

## Enconding
Dado que nuestras categorías actualmente son cadenas de texto ('Medium', 'Cheap', 'Expensive'), necesitamos codificarlas para entrenar el modelo SVM. La codificación resultante será la siguiente:

- `Expensive`: 1
- `Medium`: 2
- `Cheap`: 0

In [23]:
from sklearn.preprocessing import LabelEncoder

# Inicializar el codificador
label_encoder = LabelEncoder()

# Aplicar el Label Encoding a la columna 'PriceCategory' en df_train
df_train['PriceCategoryEncoded'] = label_encoder.fit_transform(df_train['PriceCategory'])

# Ver los valores únicos y su correspondiente encoding
print(df_train[['PriceCategory', 'PriceCategoryEncoded']].drop_duplicates())


  PriceCategory  PriceCategoryEncoded
0        Medium                     2
3         Cheap                     0
6     Expensive                     1


## Modelo de Clasificacion SVM

In [24]:
# Separar características y variable objetivo
X_train = df_train.drop(['SalePrice', 'PriceCategoryEncoded'], axis=1)
y_train = df_train['PriceCategoryEncoded']

# Manejar valores nulos (rellenar o eliminar)
X_train.fillna(0, inplace=True)  # Por ejemplo, llenar con ceros

# Convertir variables categóricas en variables numéricas (codificación one-hot)
X_train = pd.get_dummies(X_train)

# Dividir los datos en conjuntos de entrenamiento y prueba
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_train, y_train, test_size=0.2, random_state=42)


In [25]:
# Feature scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# SVM con 3 distintos modelos
svm_models = {
    'linear': SVC(kernel='linear'),
    'rbf': SVC(kernel='rbf', gamma='scale'),
    'poly': SVC(kernel='poly', degree=3)
}

# Entrenar y evaluar los modelos
for kernel, model in svm_models.items():
    model.fit(X_train_scaled, y_train)
    y_pred = model.predict(X_test_scaled)
    accuracy = accuracy_score(y_test, y_pred)
    confusion = confusion_matrix(y_test, y_pred)
    print(f"Kernel: {kernel}")
    print(f"Accuracy: {accuracy}")
    print(f"Confusion Matrix:\n{confusion}\n")

Kernel: linear
Accuracy: 1.0
Confusion Matrix:
[[139   0   0]
 [  0  24   0]
 [  0   0 129]]

Kernel: rbf
Accuracy: 0.9657534246575342
Confusion Matrix:
[[135   0   4]
 [  0  21   3]
 [  3   0 126]]

Kernel: poly
Accuracy: 0.6883561643835616
Confusion Matrix:
[[ 61   0  78]
 [  0  12  12]
 [  1   0 128]]



## Resultados SVM
- El kernel lineal logró una precisión perfecta, clasificando correctamente todas las instancias en sus respectivas categorías de precios. Esto significa que no hubo errores de clasificación, lo que indica que el modelo pudo discernir de manera óptima entre las clases 'Cheap', 'Medium' y 'Expensive'. Este resultado es altamente satisfactorio y demuestra la capacidad del modelo para generalizar bien los patrones en los datos de entrada.

- El kernel RBF también arrojó resultados muy buenos, con una precisión del 96.6%. Sin embargo, hubo algunas instancias mal clasificadas, especialmente en la categoría 'Cheap' donde se clasificaron incorrectamente 4 instancias como 'Expensive'. Además, hubo 3 instancias clasificadas incorrectamente como 'Medium' cuando en realidad eran 'Expensive'. Aunque estos errores son mínimos en comparación con el tamaño total del conjunto de datos, siguen siendo importantes ya que pueden afectar las decisiones basadas en las predicciones del modelo.

- El kernel polinomial mostró la precisión más baja de los tres modelos, con un 68.8%. Este modelo tuvo dificultades para distinguir entre las clases 'Cheap', 'Medium' y 'Expensive', como se evidencia en la matriz de confusión. Notablemente, clasificó incorrectamente 78 instancias de la clase 'Medium' como 'Expensive' y 12 instancias de la clase 'Expensive' como 'Medium'. Estos errores son significativos y podrían ser problemáticos en aplicaciones prácticas donde la precisión es crucial.

## Otro Modelo de Clasificacio Random Forest

In [28]:
# Crear el modelo de Random Forest
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)

# Entrenar el modelo
rf_model.fit(X_train_scaled, y_train)

# Predecir sobre el conjunto de prueba
y_pred = rf_model.predict(X_test_scaled)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy del modelo Random Forest:", accuracy)

# Calcular la matriz de confusión
conf_matrix = confusion_matrix(y_test, y_pred)
print("Matriz de Confusión:")
print(conf_matrix)

Accuracy del modelo Random Forest: 0.9965753424657534
Matriz de Confusión:
[[139   0   0]
 [  0  23   1]
 [  0   0 129]]


## Resultados de Random Forest
- Accuracy: El modelo de Random Forest tiene una precisión de aproximadamente 0.997, lo que significa que clasifica correctamente el 99.7% de las muestras en el conjunto de prueba.

- Matriz de Confusión: La matriz de confusión muestra que el modelo hace predicciones casi perfectas en todas las categorías, con solo una muestra clasificada incorrectamente en la categoría "Expensive".


`Para observar los resultados y conclusiones observar el pdf. `